tests/sub_group/sub_group_api.cpp

```diff
@@ -68,34 +68,106 @@ static size_t get_work_group_size(const sycl::device& device) {
+    while (tmp >>= 1) dim_size_exp++;
+    return dim_size_exp;
+  };
+
+  std::transform(desired_work_group_sizes.begin(),
+                 desired_work_group_sizes.end(),
+                 desired_work_group_sizes_exp.begin(), calc_ilog2);
+
+  auto tot_work_items = [](std::vector<size_t> sizes) {
+    return sizes[0] * sizes[1] * sizes[2];
+  };
+  std::vector<size_t> actual_work_group_sizes{
+      1, 1, static_cast<size_t>(1) << desired_work_group_sizes_exp[2]};
+
+  auto increase_dim_size = [&tot_work_items, max_work_group_size](
+                               std::vector<size_t>& wg_sizes, size_t dim_idx,
+                               size_t dim_exp) {
+    for (size_t curr_exp = 1; curr_exp <= dim_exp; curr_exp++) {
+      wg_sizes[dim_idx] <<= 1;
+      if (tot_work_items(wg_sizes) > max_work_group_size) {
+        wg_sizes[dim_idx] >>= 1;
+        break;
+      }
     }
-  }
+  };
+
+  increase_dim_size(actual_work_group_sizes, 1,
+                    desired_work_group_sizes_exp[1]);
+  increase_dim_size(actual_work_group_sizes, 0,
+                    desired_work_group_sizes_exp[0]);
+
+  return actual_work_group_sizes;
+}
 
-  return size;
+/** Returns maximum size of buffer keeping test results that can be allocated
+ * on device. */
+inline uint64_t max_device_buf_size() {
+  using buf_el_type = size_t;
+  sycl::device device = sycl_cts::util::get_cts_object::device();
+  uint64_t mem_aloc_size_in_bytes =
+      device.get_info<sycl::info::device::max_mem_alloc_size>();
+  uint64_t max_buf_size_in_elements =
+      mem_aloc_size_in_bytes / sizeof(buf_el_type);
+  return max_buf_size_in_elements;
 }
 
+/** Returns required size for buffer keeping test results. */
+inline size_t result_buf_size(size_t linear_execution_range) {
+  return linear_execution_range * member_function_cnt;
+}
+
+/** Checks that result buffer can't be allocated on device. */
+inline bool result_buffer_cant_be_alloc_on_device(size_t required_buffer_size) {
+  return required_buffer_size > max_device_buf_size();
+};
+
+/** Returns global range for given number of work group and work group sizes. */
+inline sycl::range<3> make_global_range(
+    const sycl::range<3>& work_group_num,
+    std::vector<size_t>& work_group_size_dims) {
+  return work_group_num * sycl::range<3>{work_group_size_dims[0],
+                                         work_group_size_dims[1],
+                                         work_group_size_dims[2]};
+}
+
+/**
+ * Reduces current size of each dimension of a three dimensional work-group to
+ * nearest suitable power-of-two value to ensure test result buffer will be able
+ * to be allocated on device. Returns resulting sizes for each dimension. */
+std::vector<size_t> reduce_work_group_size_to_suitable_power_of_two(
+    const sycl::range<3>& work_group_num,
+    std::vector<size_t> current_work_group_size_dims) {
+  uint64_t max_dev_buf_size = max_device_buf_size();
+  uint64_t required_res_buffer_size = result_buf_size(
```

Comment:
I am lost with some of your naming.
I am unsure that `required` is correct. Perhaps `requested` instead?